packages/core/src/server/mutation-server-protocol-handler.ts

```diff
@@ -0,0 +1,202 @@
+    const jsonRpcClient = new JSONRPCClient((message) => transporter.send(JSON.stringify(message)));
+    this.serverAndClient = new JSONRPCServerAndClient(jsonRpcServer, jsonRpcClient);
+
+    this.serverAndClient.applyServerMiddleware(this.serverInitializedMiddleware.bind(this), this.requestCancellationMiddleware.bind(this));
+    this.setupServerMethods();
+    this.setupTransporterCallbacks();
+  }
+
+  private async serverInitializedMiddleware(next: JSONRPCServerMiddlewareNext<void>, request: JSONRPCRequest, serverParams: void) {
+    if (this.initialized && request.method === 'initialize') {
+      // Server is already initialized, only allow initialize requests once.
+      return createJSONRPCErrorResponse(
+        request.id ?? null,
+        ErrorCodes.InvalidRequest,
+        'Server is already initialized, you can only request initialization once.',
+      );
+    } else if (!this.initialized && request.method !== 'initialize') {
+      // Server is not initialized, only allow initialize requests.
+      return createJSONRPCErrorResponse(
+        request.id ?? null,
+        ErrorCodes.ServerNotInitialized,
+        'Server is not initialized, you must request initialization first.',
+      );
+    }
+
+    return next(request, serverParams);
+  }
+
+  private async requestCancellationMiddleware(next: JSONRPCServerMiddlewareNext<void>, request: JSONRPCRequest, serverParams: void) {
+    if (request.id) {
+      this.cancellableRequests.set(request.id, new AbortController());
+    }
+
+    const response = await next(request, serverParams);
+
+    if (request.id) {
+      this.cancellableRequests.delete(request.id);
+    }
+
+    return response;
+  }
+
+  private setupServerMethods(): void {
+    this.serverAndClient.addMethodAdvanced('mutate', this.runMutationTestServerMethod.bind(this));
+    this.serverAndClient.addMethod('instrument', this.runInstrumentation.bind(this));
+    this.serverAndClient.addMethod('initialize', this.initialize.bind(this));
+  }
+
+  private async initialize(params: InitializeParams): Promise<typeof InitializeResult> {
+    this.strykerOptionsOverwrite.configFile = params.configUri;
+    this.initialized = true;
+    return {};
+  }
+
+  private async runInstrumentation(params: InstrumentParams): Promise<MutantResult[]> {
+    const options: PartialStrykerOptions = {
+      ...this.strykerOptionsOverwrite,
+      ...(params.globPatterns?.length && { mutate: params.globPatterns }),
+    };
+
+    return await InstrumentMethod.runInstrumentation(options);
+  }
+
+  private async runMutationTestServerMethod(jsonRPCRequest: JSONRPCRequest): Promise<JSONRPCResponse> {
+    const params = jsonRPCRequest.params as MutateParams;
+
+    if (!jsonRPCRequest.id) {
+      throw new Error('Request id is required.');
+    }
+
+    const abortController = this.cancellableRequests.get(jsonRPCRequest.id);
+
+    if (!abortController) {
+      throw new Error('No cancellation token found for request');
```

Comment:
This shouldn't happen, but because of the way you've implemented the middlewares it is actually a requirement you have to deal with it here. I think there must be a better way, let's talk about this.